### PR TITLE
Fix MICEData SettingWithCopyWarning on mixed-type dataframes

### LINKED ISSUE #5430: MICEData update produces SettingWithCopyWarning when using a data frame with mixed types
When running MICE on a data frame that contains an int column, I get a SettingWithCopyWarning (I'm not imputing the int column; in fact in my original data it's an ID column that is not used in the imputation at all). Here's a short example (note that column `b` is type integer).

```python
import pandas as pd
import numpy as np
from statsmodels.imputation.mice import MICEData

d = pd.DataFrame({'a':[1, 2, np.nan, 4], 'b':[3, 2, 1, 4]})

print(d.dtypes)

miceData = MICEData(d)
miceData.update_all()
```

output:
```
a    float64
b      int64
dtype: object
/Users/matthewc/anaconda/envs/py37/lib/python3.7/site-packages/pandas/core/indexing.py:189: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  self._setitem_with_indexer(indexer, value)
```

If you convert column b to `float`, the warning disappears. I'm guessing this is because Pandas doesn't make a copy behind the scenes when all the columns are of type float.

It appears that the culprit is [this line](https://github.com/statsmodels/statsmodels/blob/master/statsmodels/imputation/mice.py#L398), where a chain of column indexing and location-based indexing is used. 

Also mentioned in passing in #4393.

### PR DESCRIPTION
**Pull request by mattwigway**
I believe this should fix #5430; I have an automated test that passes, at least for integer dtypes.

### CODE REVIEW

**Comment by kshedden** [comment]
Thanks, this looks good.  The CI has been broken for a long time.  I don't know when we will be able to merge things again, but I'll merge this as soon as possible.

**Comment by josef-pkt** [comment]
The python 3.4 version on travis has a MICE related failure from this PR
It might not need to be fixed if we drop 3.4 for the release.
The python 3.6 fix for conda is waiting for merge, (I just forgot, moving to a new house in 3 days)

**Comment by mattwigway** [comment]
The error on Python 3.4 is from the test I added, looks like some warnings are being produced (not, I think, the warning I was trying to catch, but something else). I'll investigate further.

**Comment by josef-pkt** [comment]
I merged #5424, so travis with python 3.6 is also green on master.
The only failures are in "allowed failures" versions, i.e. problems with dev versions of dependencies.

**Comment by mattwigway** [comment]
Hey, kinda forgot about this over Christmas, sorry. I just dug more into why the test was failing on Python 3.4 and confirmed it is unrelated to the SettingWithCopyWarning issue - there's a DeprecationWarning being raised which is

```
DeprecationWarning('pandas.core.common.is_categorical_dtype is deprecated. import from the public API: pandas.api.types.is_categorical_dtype instead',)
```

I'm assuming you're not concerned about the deprecation warning that's only showing up on Python 3.4, so I just made the test ignore deprecation warnings. I also rebased my branch to latest master.

I did put the existing master on a debugger and step through the code, and it appears that even with the existing code that the data is getting updated in spite of the SettingWithCopyWarning, but there's no guarantee that that would be true in every case, I think, so I think there could be a situation where the imputation does not work as intended with the current master because `_store_values` fails to actually store the values due to the chained indexing.

Let me know if there's anything I can do to help with the CI (if that's even still an issue), would love to see this merged so I don't have to work around it in my workflow.

File: statsmodels/imputation/tests/test_mice.py [hunk 1]
```diff
@@ -106,6 +107,26 @@ def test_default(self):
...
+        df = gendat()
+        # There need to be some ints in here for the error to be thrown
+        df['intcol'] = np.arange(len(df))
+        df['intcol'] = df.intcol.astype('int32')
+
+        miceData = mice.MICEData(df)
+
+        with pd.option_context('mode.chained_assignment', 'warn'):
+            with warnings.catch_warnings(record=True) as ws:
+                warnings.simplefilter('always')
```
Comment (thequackdaddy) [inline statsmodels/imputation/tests/test_mice.py]: would `pytest.mark.filterwarnings('ignore:pandas.core.common.is_categorical_dtype is deprecated:DeprecationWarning')` work better? you might still be `able to assert_no_warnings()`? 

You might be able to do...

```
@pytest.mark.filterwarnings('ignore:pandas.core.common.is_categorical_dtype is deprecated:DeprecationWarning')
def test_settingwithcopywarning(self):
    df = gendat()
    # There need to be some ints in here for the error to be thrown
    df['intcol'] = np.arange(len(df))
    df['intcol'] = df.intcol.astype('int32')
	
    miceData = mice.MICEData(df)
    with assert_no_warnings():
        miceData.update_all()
```

Comment (mattwigway) [reply statsmodels/imputation/tests/test_mice.py]: Good thought. I just tried that though and unfortunately it doesn't work - `assert_no_warnings` still fails (must be catching them below where pytest does its filtering).

**Pull request closed without merging**